File: internal/tasks/parser.go
// Copyright (C) 2026 Zekrost <tech@zekrost.com>
// SPDX-License-Identifier: AGPL-3.0-only
// Package tasks implementa el motor de tareas embebidas: convierte
// Markdown plano en un sistema de gestión sin bases de datos de tareas
// separadas. Es el activo intelectual central del producto.
//
// Gramática (Documento Técnico de Arquitectura, sección 6.1):
//
//	tarea := checkbox WS texto (WS metadato)*
//	checkbox := '- [ ]' | '- [x]' | '- [~]'      -- ~ = en progreso
//	metadato := fecha | proyecto | prioridad | asignado | etiqueta
//	fecha := '#' (AAAA-MM-DD | 'hoy' | 'mañana' | 'lun'..'dom')
//	proyecto := '@' ident
//	prioridad := '!' ('baja'|'media'|'alta'|'1'..'3')
//	asignado := '~' ident
//	etiqueta := '+' ident
//	ident := [a-z0-9-]+
//
// Invariantes: idempotente, tolerante y con round-trip garantizado:
// toda edición desde una vista reescribe la línea original preservando
// el resto del texto al byte (sección 6.2).
package tasks

import (
	"strings"
	"time"
)

// State de una tarea embebida.
const (
	StateOpen     = " "
	StateDone     = "x"
	StateProgress = "~"
)

// Task es la representación parseada de una línea de checkbox.
// RawLine conserva la línea original para garantizar el round-trip.
type Task struct {
	Line      int      `json:"line_no"` // 1-based dentro del documento
	RawLine   string   `json:"-"`
	Text      string   `json:"title"`
	Done      bool     `json:"done"`
	InProgress bool    `json:"in_progress"`
	DueDate   string   `json:"due_date"` // AAAA-MM-DD
	Project   string   `json:"project"`
	Priority  string   `json:"priority"` // baja | media | alta
	Assignee  string   `json:"assignee"`
	Tags      []string `json:"tags"`
}

// ParseLine intenta parsear una línea de documento como tarea embebida.
// Devuelve ok=false si la línea no es un checkbox Markdown.
func ParseLine(line string) (Task, bool) {
	trimmed := strings.TrimRight(line, "\r\n ")
	if !strings.HasPrefix(trimmed, "- [") {
		return Task{}, false
	}
	state, rest, ok := parseCheckbox(trimmed)
	if !ok {
		return Task{}, false
	}

	t := Task{
		RawLine: trimmed,
		Done:    state == StateDone,
		InProgress: state == StateProgress,
	}

	// texto + metadatos: separados por espacios simples
	parts := strings.Fields(rest)
	textParts := make([]string, 0, len(parts))
	for _, p := range parts {
		switch {
		case strings.HasPrefix(p, "#"):
			t.DueDate = parseDate(p[1:])
		case strings.HasPrefix(p, "@") && isIdent(p[1:]):
			t.Project = p[1:]
		case strings.HasPrefix(p, "!") && isPriority(p[1:]):
			t.Priority = normalizePriority(p[1:])
		case strings.HasPrefix(p, "~") && isIdent(p[1:]):
			t.Assignee = p[1:]
		case strings.HasPrefix(p, "+") && isIdent(p[1:]):
			t.Tags = append(t.Tags, p[1:])
		default:
			// tolerante: metadatos desconocidos se conservan como texto
			textParts = append(textParts, p)
		}
	}
	t.Text = strings.Join(textParts, " ")
	return t, true
}

// Parse recorre un documento y extrae todas las tareas embebidas.
// Idempotente: parsear dos veces el mismo documento produce el mismo
// conjunto (clave: doc_id + line_no + hash de línea).
func Parse(content string) []Task {
	var out []Task
	for i, line := range strings.Split(content, "\n") {
		if t, ok := ParseLine(line); ok {
			t.Line = i + 1
			out = append(out, t)
		}
	}
	return out
}

// RoundTrip reescribe la línea original de una tarea tras una mutación,
// preservando al byte el texto que no sea de metadatos. Garantiza que
// completar/reprogramar/reasignar desde una vista nunca corrompa la
// redacción original (sección 6.2).
func RoundTrip(t Task, done bool, due, project, priority, assignee string) string {
	fields := strings.Fields(t.RawLine)
	var out []string
	// conserva el estado original del checkbox (o lo actualiza)
	head := "- [" + stateOf(t, done) + "]"
	out = append(out, head)
	// "- [ ]" se parte en 3 tokens al hacer Fields; se saltan
	rest := fields
	if len(rest) >= 3 && rest[0] == "-" && rest[1] == "[" && strings.HasSuffix(rest[2], "]") {
		rest = rest[3:]
	}
	wroteText := false
	for _, p := range rest {
		switch {
		case strings.HasPrefix(p, "#"):
			if due != "" {
				out = append(out, "#"+due)
			} else {
				out = append(out, p) // sin cambio: se conserva el token original
			}
		case strings.HasPrefix(p, "@"):
			if project != "" {
				out = append(out, "@"+project)
			} else {
				out = append(out, p)
			}
		case strings.HasPrefix(p, "!"):
			if priority != "" {
				out = append(out, "!"+priority)
			} else {
				out = append(out, p)
			}
		case strings.HasPrefix(p, "~"):
			if assignee != "" {
				out = append(out, "~"+assignee)
			} else {
				out = append(out, p)
			}
		default:
			// texto: se vuelve a escribir tal cual, al byte
			if !wroteText {
				out = append(out, strings.TrimSpace(p))
				wroteText = true
			} else {
				out = append(out, p)
			}
		}
	}
	return strings.Join(out, " ")
}

func stateOf(t Task, done bool) string {
	if done {
		return StateDone
	}
	if t.InProgress {
		return StateProgress
	}
	return StateOpen
}

func parseCheckbox(line string) (state, rest string, ok bool) {
	// "- [ ]" | "- [x]" | "- [~]" — exige espacio tras ']'
	if len(line) < 6 || line[1] != ' ' || line[2] != '[' || line[4] != ']' || line[5] != ' ' {
		return "", "", false
	}
	c := line[3]
	switch c {
	case ' ', 'x', '~':
	default:
		return "", "", false
	}
	rest = strings.TrimSpace(line[5:])
	if rest == "" {
		return "", "", false // tarea sin texto: se ignora
	}
	return string(c), rest, true
}

func isIdent(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if !(r >= 'a' && r <= 'z' || r >= '0' && r <= '9' || r == '-') {
			return false
		}
	}
	return true
}

func isPriority(s string) bool {
	switch s {
	case "baja", "media", "alta", "1", "2", "3":
		return true
	}
	return false
}

func normalizePriority(s string) string {
	switch s {
	case "1":
		return "alta"
	case "2":
		return "media"
	case "3":
		return "baja"
	}
	return s
}

// parseDate resuelve fechas relativas ('hoy', 'mañana', 'lun'..'dom')
// a AAAA-MM-DD; fechas inválidas devuelven "" sin romper el parseo
// (regla tolerante, sección 6.2).
func parseDate(raw string) string {
	if raw == "" {
		return ""
	}
	if len(raw) == 10 && raw[4] == '-' && raw[7] == '-' {
		return raw // AAAA-MM-DD válida o no: se conserva
	}
	now := time.Now()
	switch strings.ToLower(raw) {
	case "hoy":
		return now.Format("2006-01-02")
	case "mañana":
		return now.AddDate(0, 0, 1).Format("2006-01-02")
	}
	dow := map[string]time.Weekday{
		"lun": time.Monday, "mar": time.Tuesday, "mie": time.Wednesday,
		"jue": time.Thursday, "vie": time.Friday, "sab": time.Saturday,
		"dom": time.Sunday,
	}
	if wd, ok := dow[strings.ToLower(raw)]; ok {
		delta := (int(wd) - int(now.Weekday()) + 7) % 7
		if delta == 0 {
			delta = 7 // el próximo día con ese nombre
		}
		return now.AddDate(0, 0, delta).Format("2006-01-02")
	}
	return "" // fecha inválida: warning en la UI, nunca rompe el parseo
}

File: web/src/tasks/parser.ts
// Port del parser de tareas embebidas del backend Go
// (internal/tasks/parser.go) — misma gramática y reglas.
//
//	tarea := checkbox WS texto (WS metadato)*
//	checkbox := '- [ ]' | '- [x]' | '- [~]'      -- ~ = en progreso
//	metadato := fecha | proyecto | prioridad | asignado | etiqueta
//	fecha := '#' (AAAA-MM-DD | 'hoy' | 'mañana' | 'lun'..'dom')
//	proyecto := '@' ident   prioridad := '!' (baja|media|alta|1..3)
//	asignado := '~' ident   etiqueta := '+' ident
//	ident := [a-z0-9-]+
//
// Invariantes: idempotente, tolerante y round-trip garantizado (la
// edición desde una vista reescribe la línea preservando el resto).

export interface ParsedTask {
  line: number; // 1-based
  rawLine: string;
  title: string;
  done: boolean;
  inProgress: boolean;
  dueDate: string | null; // AAAA-MM-DD
  project: string | null;
  priority: string | null; // baja | media | alta
  assignee: string | null;
  tags: string[];
}

export function parseLine(line: string): ParsedTask | null {
  const trimmed = line.replace(/[ \t]+$/, "");
  // checkbox estricto: '- [ ]' requiere espacio tras ']'
  const m = /^-\s\[([ xX~])\]\s+(.+)$/.exec(trimmed);
  if (!m) return null;
  const state = m[1].toLowerCase();
  const rest = m[2];

  const t: ParsedTask = {
    line: 0,
    rawLine: trimmed,
    done: state === "x",
    inProgress: state === "~",
    title: "",
    dueDate: null,
    project: null,
    priority: null,
    assignee: null,
    tags: [],
  };

  const textParts: string[] = [];
  for (const p of rest.split(/\s+/)) {
    if (p.startsWith("#")) {
      const d = resolveDateISO(p.slice(1));
      if (d) t.dueDate = d;
      else textParts.push(p);
    } else if (p.startsWith("@") && isIdent(p.slice(1))) {
      t.project = p.slice(1);
    } else if (p.startsWith("!") && isPriority(p.slice(1))) {
      t.priority = normalizePriority(p.slice(1));
    } else if (p.startsWith("~") && isIdent(p.slice(1))) {
      t.assignee = p.slice(1);
    } else if (p.startsWith("+") && isIdent(p.slice(1))) {
      t.tags.push(p.slice(1));
    } else {
      textParts.push(p); // tolerante: metadatos desconocidos se conservan
    }
  }
  t.title = textParts.join(" ");
  return t;
}

export function parse(content: string): ParsedTask[] {
  const out: ParsedTask[] = [];
  const lines = content.split("\n");
  for (let i = 0; i < lines.length; i++) {
    const t = parseLine(lines[i]);
    if (t) {
      t.line = i + 1;
      out.push(t);
    }
  }
  return out;
}

// RoundTrip reescribe la línea original tras una mutación, preservando
// al byte el texto que no es metadato.
export function roundTrip(
  t: ParsedTask,
  done: boolean,
  due?: string | null,
  project?: string | null,
  priority?: string | null,
  assignee?: string | null,
): string {
  const fields = t.rawLine.split(/\s+/);
  const head = "- [" + (done ? "x" : t.inProgress ? "~" : " ") + "]";
  const out: string[] = [head];
  // '- [ ]' se parte en 3 tokens al hacer split
  const rest = fields.slice(3);
  let wroteText = false;
  for (const p of rest) {
    if (p.startsWith("#")) {
      if (due) out.push("#" + due);
      else out.push(p);
    } else if (p.startsWith("@")) {
      if (project) out.push("@" + project);
      else out.push(p);
    } else if (p.startsWith("!")) {
      if (priority) out.push("!" + priority);
      else out.push(p);
    } else if (p.startsWith("~")) {
      if (assignee) out.push("~" + assignee);
      else out.push(p);
    } else {
      if (!wroteText) {
        out.push(p.trim());
        wroteText = true;
      } else {
        out.push(p);
      }
    }
  }
  return out.join(" ");
}

// Aplica un cambio de estado a un documento completo: reescribe la línea
// de la tarea y devuelve el contenido nuevo.
export function applyTaskState(content: string, task: ParsedTask, done: boolean): string {
  const lines = content.split("\n");
  const idx = task.line - 1;
  if (idx < 0 || idx >= lines.length) return content;
  lines[idx] = roundTrip(task, done);
  return lines.join("\n");
}

function isIdent(s: string): boolean {
  return /^[a-z0-9-]+$/.test(s);
}

function isPriority(s: string): boolean {
  return /^(baja|media|alta|1|2|3)$/.test(s);
}

function normalizePriority(s: string): string {
  switch (s) {
    case "1":
      return "alta";
    case "2":
      return "media";
    case "3":
      return "baja";
  }
  return s;
}

// Resuelve fechas relativas a AAAA-MM-DD. Devuelve null si no es fecha.
export function resolveDateISO(raw: string): string | null {
  if (/^\d{4}-\d{2}-\d{2}$/.test(raw)) return raw;
  const now = new Date();
  const iso = (d: Date) => d.toISOString().split("T")[0];
  switch (raw.toLowerCase()) {
    case "hoy":
      return iso(now);
    case "mañana":
      return iso(new Date(now.getTime() + 86400000));
  }
  const days = ["lunes", "martes", "miércoles", "jueves", "viernes", "sábado", "domingo"];
  const dayIdx = days.indexOf(raw.toLowerCase());
  if (dayIdx >= 0) {
    const target = dayIdx + 1; // 1..7, lunes=1
    const cur = now.getDay() === 0 ? 7 : now.getDay(); // domingo=7
    let diff = target - cur;
    if (diff <= 0) diff += 7;
    const d = new Date(now);
    d.setDate(d.getDate() + diff);
    return iso(d);
  }
  return null;
}
